File: events.py
"""Events tab — next scheduled catalyst per contract."""
import calendar
import datetime as _d

import pandas as pd
import streamlit as st


def _next_weekday(wd: int) -> _d.date:
    """Next date (today or later) falling on weekday wd (Mon=0 .. Sun=6)."""
    t = _d.date.today()
    return t + _d.timedelta(days=(wd - t.weekday()) % 7)


def _first_friday(y: int, m: int) -> _d.date:
    d = _d.date(y, m, 1)
    return d + _d.timedelta(days=(4 - d.weekday()) % 7)


def _next_first_friday() -> _d.date:
    t = _d.date.today()
    ff = _first_friday(t.year, t.month)
    if ff < t:
        y, m = (t.year + 1, 1) if t.month == 12 else (t.year, t.month + 1)
        ff = _first_friday(y, m)
    return ff


def _next_from_list(dates: list) -> _d.date | None:
    t = _d.date.today()
    upcoming = [d for d in dates if d >= t]
    return min(upcoming) if upcoming else None


def _last_business_day(y: int, m: int) -> _d.date:
    last = calendar.monthrange(y, m)[1]
    d = _d.date(y, m, last)
    while d.weekday() > 4:
        d -= _d.timedelta(days=1)
    return d


def _next_month_day(day: int) -> _d.date:
    """Next occurrence of the given day-of-month (today or future)."""
    t = _d.date.today()
    cand = _d.date(t.year, t.month, min(day, calendar.monthrange(t.year, t.month)[1]))
    if cand < t:
        y, m = (t.year + 1, 1) if t.month == 12 else (t.year, t.month + 1)
        cand = _d.date(y, m, min(day, calendar.monthrange(y, m)[1]))
    return cand


# Confirmed 2026 fixed dates
FOMC_2026 = [_d.date(2026, 7, 29), _d.date(2026, 9, 16),
             _d.date(2026, 10, 28), _d.date(2026, 12, 9)]
CPI_2026 = [_d.date(2026, 8, 12), _d.date(2026, 9, 11),   # confirmed
            _d.date(2026, 10, 13), _d.date(2026, 11, 12), _d.date(2026, 12, 10)]  # ≈

# name, next-date callable, time (ET), impact, affected instruments, exact?
EVENTS = [
    ("EIA Petroleum Status", lambda: _next_weekday(2), "10:30", "High", ["CL"], True),
    ("EIA Nat Gas Storage",  lambda: _next_weekday(3), "10:30", "High", ["NG"], True),
    ("API Crude (private)",  lambda: _next_weekday(1), "16:30", "Med",  ["CL"], True),
    ("Nonfarm Payrolls",     _next_first_friday,       "08:30", "High", ["ES", "NQ", "GC", "SI", "6E", "6J"], True),
    ("Jobless Claims",       lambda: _next_weekday(3), "08:30", "Med",  ["ES", "NQ"], True),
    ("FOMC Rate Decision",   lambda: _next_from_list(FOMC_2026), "14:00", "High", ["ES", "NQ", "GC", "SI", "HG", "6E", "6J"], True),
    ("CPI Inflation",        lambda: _next_from_list(CPI_2026),  "08:30", "High", ["ES", "NQ", "GC", "SI", "6E", "6J"], True),
    ("PCE Inflation",        lambda: _last_business_day(_d.date.today().year, _d.date.today().month), "08:30", "High", ["ES", "NQ", "GC", "SI"], False),
    ("USDA WASDE",           lambda: _next_month_day(12), "12:00", "High", ["ZC", "ZS"], False),
    ("USDA Crop Progress",   lambda: _next_weekday(0), "16:00", "Med",  ["ZC", "ZS"], False),
    ("USDA Export Sales",    lambda: _next_weekday(3), "08:30", "Med",  ["ZC", "ZS"], True),
]

INSTRUMENTS = ["ES", "NQ", "GC", "SI", "HG", "CL", "NG", "ZC", "ZS", "6E", "6J"]


def build_events(selected: list) -> pd.DataFrame:
    t = _d.date.today()
    rows = []
    for name, fn, time_et, impact, insts, exact in EVENTS:
        if selected and not (set(insts) & set(selected)):
            continue
        try:
            d = fn()
        except Exception:  # noqa: BLE001
            d = None
        if d is None:
            continue
        days = (d - t).days
        when = "today" if days == 0 else "tomorrow" if days == 1 else f"in {days}d"
        rows.append({
            "_sort": d,
            "Date": ("≈ " if not exact else "") + d.strftime("%a %d %b"),
            "Time ET": time_et,
            "Event": name,
            "Impact": impact,
            "Affects": " ".join(insts),
            "Countdown": when,
        })
    df = pd.DataFrame(rows).sort_values("_sort")
    return df.drop(columns="_sort") if not df.empty else df


def render_events() -> None:
    st.caption(
        "Next scheduled catalyst per contract. ✓ dates are rule-based or "
        "confirmed (auto-rolling); **≈** dates are estimates — verify before "
        "trading. Euro/Yen also move on ECB/BOJ decisions (check their calendars)."
    )
    sel = st.multiselect("Filter by instrument", INSTRUMENTS, default=[])
    df = build_events(sel)
    if df.empty:
        st.info("No upcoming events for that selection.")
    else:
        st.table(df.style.hide(axis="index"))

File: site/js/events.js
/* events.js — computed in the browser from calendar rules alone.

   Deliberately not built into JSON. Roll dates and expiries follow fixed
   rules, so computing them at page load keeps them correct even when the last
   build is days old — the one tab that cannot go stale. */

function nextWeekday(wd) {
  var t = new Date(); t.setHours(0, 0, 0, 0);
  var d = new Date(t);
  d.setDate(t.getDate() + ((wd - ((t.getDay() + 6) % 7)) + 7) % 7);
  return d;
}

function firstFriday(y, m) {
  var d = new Date(y, m, 1);
  d.setDate(1 + ((5 - d.getDay()) + 7) % 7);
  return d;
}

function nextFirstFriday() {
  var t = new Date(); t.setHours(0, 0, 0, 0);
  var f = firstFriday(t.getFullYear(), t.getMonth());
  if (f < t) f = firstFriday(t.getFullYear(), t.getMonth() + 1);
  return f;
}

function lastBusinessDay() {
  var t = new Date();
  var d = new Date(t.getFullYear(), t.getMonth() + 1, 0);
  while (d.getDay() === 0 || d.getDay() === 6) d.setDate(d.getDate() - 1);
  if (d < new Date(t.getFullYear(), t.getMonth(), t.getDate())) {
    d = new Date(t.getFullYear(), t.getMonth() + 2, 0);
    while (d.getDay() === 0 || d.getDay() === 6) d.setDate(d.getDate() - 1);
  }
  return d;
}

function nextMonthDay(day) {
  var t = new Date(); t.setHours(0, 0, 0, 0);
  var d = new Date(t.getFullYear(), t.getMonth(), day);
  if (d < t) d = new Date(t.getFullYear(), t.getMonth() + 1, day);
  return d;
}

function nextFrom(list) {
  var t = new Date(); t.setHours(0, 0, 0, 0);
  var up = list.map(function (s) { return new Date(s + "T00:00:00"); })
    .filter(function (d) { return d >= t; });
  return up.length ? up[0] : null;
}

var FOMC = ["2026-09-16", "2026-10-28", "2026-12-09", "2027-01-27"];

var CPI = ["2026-08-12", "2026-09-11", "2026-10-13", "2026-11-12", "2026-12-10"];

var EVENTS = [
  ["EIA Petroleum Status", function () { return nextWeekday(2); }, "10:30", "High", ["CL"], true],
  ["EIA Nat Gas Storage", function () { return nextWeekday(3); }, "10:30", "High", ["NG"], true],
  ["API Crude (private)", function () { return nextWeekday(1); }, "16:30", "Med", ["CL"], true],
  ["Nonfarm Payrolls", nextFirstFriday, "08:30", "High", ["ES", "NQ", "NKD", "GC", "SI", "6E", "6J"], true],
  ["Jobless Claims", function () { return nextWeekday(3); }, "08:30", "Med", ["ES", "NQ"], true],
  ["FOMC Rate Decision", function () { return nextFrom(FOMC); }, "14:00", "High", ["ES", "NQ", "GC", "SI", "HG", "6E", "6J"], true],
  ["CPI Inflation", function () { return nextFrom(CPI); }, "08:30", "High", ["ES", "NQ", "GC", "SI", "6E", "6J"], true],
  ["PCE Inflation", lastBusinessDay, "08:30", "High", ["ES", "NQ", "GC", "SI"], false],
  ["USDA WASDE", function () { return nextMonthDay(12); }, "12:00", "High", ["ZC", "ZS", "ZW"], false],
  ["USDA Crop Progress", function () { return nextWeekday(0); }, "16:00", "Med", ["ZC", "ZS"], false],
  ["USDA Export Sales", function () { return nextWeekday(3); }, "08:30", "Med", ["ZC", "ZS", "ZW"], true]
];

function renderEvents() {
  var t = new Date(); t.setHours(0, 0, 0, 0);
  var filt = S.events.filter;
  var rows = EVENTS.map(function (e) {
    var d = null;
    try { d = e[1](); } catch (x) { d = null; }
    if (!d) return null;
    if (filt !== "All" && e[4].indexOf(filt) < 0) return null;
    var days = Math.round((d - t) / 86400000);
    return {
      d: d, when: days === 0 ? "today" : days === 1 ? "tomorrow" : "in " + days + "d",
      name: e[0], time: e[2], impact: e[3], affects: e[4].join(" "), exact: e[5]
    };
  }).filter(Boolean).sort(function (a, b) { return a.d - b.d; });

  var codes = ["All"].concat(META.universe.map(function (u) { return u.code; })
    .filter(function (c) {
      return EVENTS.some(function (e) { return e[4].indexOf(c) >= 0; });
    }));
  var head = '<th class="l">Date</th><th class="l">Time ET</th>' +
    '<th class="l">Event</th><th class="l">Impact</th><th class="l">Affects</th>' +
    '<th class="l">Countdown</th>';
  var body = rows.map(function (r) {
    var ds = r.d.toDateString().slice(0, 10);
    return '<tr><td class="l">' + (r.exact ? "" : "≈ ") + esc(ds) +
      '</td><td class="l dim">' + r.time + '</td><td class="l">' + esc(r.name) +
      '</td><td class="l ' + (r.impact === "High" ? "neg" : "dim") + '">' +
      r.impact + '</td><td class="l faint">' + esc(r.affects) +
      '</td><td class="l dim">' + r.when + "</td></tr>";
  }).join("");

  view(
    '<div class="note">Next scheduled catalyst per contract, computed in the ' +
    "browser from calendar rules — so these stay correct even if the data " +
    "build is a day stale. <b>≈</b> marks an estimate; verify before trading. " +
    "Euro and Yen also move on ECB and BOJ decisions, which are not on this " +
    "list.</div>" +
    '<div class="bar"><select data-sel="evFilter">' + codes.map(function (c) {
      return '<option' + (c === filt ? " selected" : "") + ">" + c + "</option>";
    }).join("") + "</select></div>" +
    (rows.length ? table(head, body)
      : '<div class="skel">No upcoming events for that selection.</div>')
  );
}

/* --------------------------------------------------------------- Drivers */
/* A standing frame, not a feed. These change when the structure of a market
   changes, which is why they live in the repo rather than in a scrape. */
